### ISSUE TITLE
Add unit tests

### ISSUE THREAD

**Issue by vsiakka** [vsiakka/gherkin-lint#7]

**Comment by timbru31** [comment]
In case you don't know, once the tests are finished, you could enable https://greenkeeper.io/ for this repo.
The service monitors your dependency and ranges, once an update is released it creates a PR with the dependency update and if Travis is configured the tests will run with the associated PR 🎉

**Comment by vsiakka** [comment]
That's great! I will take a look :)

**Comment by nene** [comment]
I'd like to implement some unit tests for the new rules I've done pull requests for, but I found the structure of unit tests in `test/rules/` dir pretty hard to follow.

What I'd like to see:

- simple tests checking that given a specific input produces expected error.

What I currently see:

- `test-data/` files with good chunk of code that either generate no errors, or generates a bunch of errors
- `results.js` file defining all the expected errors, grouped by some names.
- `test.js` file that maps test-data files to expected errors. Reading in files, running the Gherkin parser, executing the rule - quite a lot of boilerplate for each test. It requires quite a bit of mental juggling to see how the test-data files correspond to the expected errors.

What I would propose:

- a) Writing the feature file string directly inside the test and writing the expected errors inside there as well. This way one can write small tests that check for a single error at a time. One should also write a helper function that can be used in all tests instead of reimplementing the logic in each test suite.
- b) Having one errors file for each test-data file. For example by having `TestName.feature` and `TestName.errors` side-by-side. One generic test would then parse all these files and perform comparison of expected results - no need to the mocha tests, adding the feature and error files should be enough. Testing the different config options looks more cumbersome with this approach, therefore I would rather prefer the former.

**Comment by vsiakka** [comment]
I hear you, the unit tests are not in a good state, it's my first time writing unit tests so i wanted to write a few and then adjust to w/e made sense. I want to think about the layout a little bit more when i get some time and then I will reply with something more related to your suggesting @nene .

**Comment by xgbuils** [comment]
Hi,

Currently this project only has functional tests. In my opinion unit tests should test small pieces of code and it is not necessary to import files. For example in `use-and`. We can extract a function that checks if there are error in steps of one scenario:

```
  function checkUseAndInScenario(steps) {
    var errors = [];
    var previousKeyword = undefined;
    steps.forEach(function(step) {
      if (step.keyword === 'And ') {
        return;
      }
      if (step.keyword === previousKeyword) {
        errors.push(createError(step));
      }
      previousKeyword = step.keyword;
    });
    return errors;
  }

// ...
function useAnd(feature) {
  var errors = [];

  feature.children.forEach(function(child) {
    [].push.apply(errors, checkUseAndInScenario(child.steps))
  });

  return errors;
}
```

And then, we can test `checkUseAndInScenario` with some inputs like:

- Repeated Given:
`steps = [{keyword: 'Given'}, {keyword: 'Given'}]`
- Repeated When:
`steps = [{keyword: 'When'}, {keyword: 'When'}]`
- Repeated Then, And, Then:
`steps = [{keyword: 'Then'}, {keyword: 'And'}, {keyword: 'Then'}]`
- Repeated And:
`steps = [{keyword: 'When'}, {keyword: 'And'}, {keyword: 'And'}]`
- Starting with And:
`steps = [{keyword: 'When'}, {keyword: 'And'}, {keyword: 'And'}]`
- One step Given:
`steps = [{keyword: 'Given'}]`
- One step And:
`steps = [{keyword: 'And'}]`
- Correct steps without And:
`steps = [{keyword: 'Given'}, {keyword: 'When'}, {keyword: 'Then'}, , {keyword: 'When'}, {keyword: 'Then'}]`
- Correct steps with And:
`steps = [{keyword: 'Given'}, {keyword: 'And'}, {keyword: 'When'}, {keyword: 'And'}, {keyword: 'Then'}, {keyword: 'And'}]`
- Empty steps:
`steps = []`

and check if  `checkUseAndInScenario(steps)` returns the expected array of errors.

IMHO it is absolutely important to start doing better unit tests and adding tests for each piece of logic if we can provide a reliable package. For example, without unit tests we cannot detect that `use-and` rule does not report error in the following step sequency: `Given, And, Given, And, When, And, When, And, Then, And, Then`

**Comment by vsiakka** [comment]
Test coverage is high enough that I feel comfortable closing this. Coveralls is also configured to fail builds that drop coverage % significantly.
![image](https://user-images.githubusercontent.com/14342371/57186665-1b3f9600-6eec-11e9-8254-6c81ac94db1f.png)
